### ISSUE TITLE
3.1.0: Changes in Smart.Default don't work in other threads

### ISSUE THREAD

**Issue by za9csergei** [axuno/SmartFormat#287]
**Smart.Format version**:  3.1.0

**Framework version**:  net 6.0

In asp.net 6 application in Program.cs I change Smart.Default to ignore Formatter and Parser errors, enable StringFormatCompatibility, add IsMatchFormatter with RegexOptions = RegexOptions.CultureInvariant. But these settings work only in Program file, but don't work in other file. For example in controllers. And IsMatchFormatter doesn't work even in Program. 

- Please provide a working source code example to reproduce the bug
I made a reproducible demo. Asp.net 6 app with Program file, TestController with Test method and workaround static calss CustomFormatter https://github.com/za9csergei/smartformat3setupdemo

Program.cs

```csharp
Smart.Default.Settings.Formatter.ErrorAction = FormatErrorAction.Ignore;
Smart.Default.Settings.Parser.ErrorAction = ParseErrorAction.Ignore;

// don't apply changes in TestController too
//Smart.Default = Smart.CreateDefaultSmartFormat(new SmartSettings
//{
//    Formatter = new FormatterSettings { ErrorAction = FormatErrorAction.Ignore },
//    Parser = new ParserSettings { ErrorAction = ParseErrorAction.Ignore }
//});

Smart.Default.Settings.StringFormatCompatibility = true;

Smart.Default.AddExtensions(new TimeFormatter());
Smart.Default.AddExtensions(new IsMatchFormatter { RegexOptions = RegexOptions.CultureInvariant });

var test = new
{
    CreateDate = new DateTime(2022, 5, 22, 12, 7, 33),
    Name = "John",
    Street = "4th street"
};

var str = Smart.Format("{CreateDate:dd.MM.yyyy HH:mm} {Name} {Street:ismatch(th):TH|th}", test);
// expected : 22.05.2022 12:07 John TH
// actual:    22.05.2022 12:07 John 4th street
```

TestController.Test
```csharp
public string Test()
{
    var test = new
    {
        CreateDate = new DateTime(2022, 5, 22, 12, 7, 33),
        Name = "John",
        Street = "4th street"
    };

    var customFormatResult = CustomFormat.Format("{CreateDate:dd.MM.yyyy HH:mm} {Name} {Street:ismatch(th):TH|th}",
        test
    );
    Console.WriteLine(customFormatResult); // 22.05.2022 12:07 John 4th street

    var smartFormatResult = Smart.Format("{CreateDate:dd.MM.yyyy HH:mm} {Name} {Street:ismatch(th):TH|th}",
        test
    ); // throws FormattingException: Error parsing format string: No formatter with name 'dd.MM.yyyy HH' found at 0 {CreateDate:dd.MM.yyyy HH:mm} {Name} {Street:ismatch(th):TH|th}

    return smartFormatResult;
}
```

- What is the current result?
  - Program Smart.Format call ismatch ignored - 22.05.2022 12:07 John 4th street
  - TestController.Test Smart.Format call throws exception - FormattingException: Error parsing format string: No formatter with name 'dd.MM.yyyy HH' found at 0 {CreateDate:dd.MM.yyyy HH:mm} {Name} {Street:ismatch(th):TH|th}
- What is the expected result?
Both Smart.Format calls should return string "22.05.2022 12:07 John TH"

- Please post full exception details if applicable (message, stacktrace, inner exceptions)
```json
{
  "ClassName": "SmartFormat.Core.Formatting.FormattingException",
  "Message": null,
  "Data": null,
  "InnerException": null,
  "HelpURL": null,
  "StackTraceString": "   at SmartFormat.SmartFormatter.FormatError(FormatItem errorItem, Exception innerException, Int32 startIndex, IFormattingInfo formattingInfo)\r\n   at SmartFormat.SmartFormatter.Format(FormattingInfo formattingInfo)\r\n   at SmartFormat.SmartFormatter.Format(FormatDetails formatDetails, Format format, Object current)\r\n   at SmartFormat.SmartFormatter.Format(IFormatProvider provider, String format, IList`1 args)\r\n   at SmartFormat.SmartFormatter.Format(String format, Object[] args)\r\n   at SmartFormat.Smart.Format(String format, Object arg0)\r\n   at SmartFormat3SetupDemo.Controllers.TestController.Test() in C:\\Users\\SergeiZaitsev\\Documents\\smartformat3setupdemo\\SmartFormat3SetupDemo\\SmartFormat3SetupDemo\\Controllers\\TestController.cs:line 27",
  "RemoteStackTraceString": null,
  "RemoteStackIndex": 0,
  "ExceptionMethod": null,
  "HResult": -2146233088,
  "Source": "SmartFormat",
  "WatsonBuckets": null
}
```
- Did you find a workaround? 
Yes. I created a static class. I initialized in It SmartFormatter instance ```_formatter``` with prefered settings and method Format that calls _formatter.Format. Settings, except IsMatchFormatter, work in any file
```csharp
public static class CustomFormat
{
    private static readonly SmartFormatter _formatter = CreateCustomFormatter();

    private static SmartFormatter CreateCustomFormatter()
    {
        var formatter = Smart.CreateDefaultSmartFormat(new SmartSettings
        {
            Formatter = new FormatterSettings { ErrorAction = FormatErrorAction.Ignore },
            Parser = new ParserSettings { ErrorAction = ParseErrorAction.Ignore }
        });

        formatter.Settings.StringFormatCompatibility = true;
        formatter.AddExtensions(new TimeFormatter());
        formatter.AddExtensions(new IsMatchFormatter { RegexOptions = RegexOptions.CultureInvariant });
        return formatter;
    }

    public static string Format(string format, params object?[] args) => _formatter.Format(format, args);
}
```

- Is there a version in which it worked? 
2.7.0. In this version was enough following code in Program to settings work in any file
```csharp
Smart.Default.Settings.FormatErrorAction = ErrorAction.Ignore;
Smart.Default.Settings.ParseErrorAction = ErrorAction.Ignore;
Smart.Default.FormatterExtensions.Add(new IsMatchFormatter { RegexOptions = RegexOptions.CultureInvariant });
```

Do I something wrong when work with Smart.Default and IsMatchFormatter in 3.1.0 version?

**Comment by axunonb** [comment]
Thanks for pointing this out.

**Multi-threading**  (assuming the controller is in Asp.Net)
`Smart.Default` is thread-static in v3.1.0. Each thread gets a new instance of the `SmartFormatter`.
That's why your initial configuration only has effect for the thread you assign `Smart.Default`. Your "workaround" is the way to go here, also for best control how the `SmartFormatter` is initialized. It's always better to just load the required extensions.

**`SmartSettings..StringFormatCompatibility = true`**
`Settings.StringFormatCompatibility = true` is a strong limitation as for *SmartFormat* features. Please have a look at the [Wiki chapter](https://github.com/axuno/SmartFormat/wiki/string.Format-Compatibility) covering this topic.
The default is `SmartSettings.StringFormatCompatibility = false`
**TLDR;** 
`var smartFormatResult = Smart.Format(@"{CreateDate:dd.MM.yyyy HH\:mm} {Name} {Street:ismatch(th):TH|th}", test);`
(with required escaped colon in 'HH\:mm' for this mode will not throw).

**Comment by za9csergei** [comment]
Thank you!

>  As long as special characters (){}:\ are escaped with \, any character is allowed anywhere

I was not carefull enough with escaping and multi-threading and when read about v3.1.0 and migration. 
Removing ```Settings.StringFormatCompatibility = true``` and escaping colon made work ```ismatch``` and DateTime formatting as expected!

**Comment by axunonb** [comment]
You're welcome. First feedback for v3.1.0 ;-)

**Comment by axunonb** [comment]
I'm currently out-of-office.
Did you already have a look at [this wiki article](https://github.com/axuno/SmartFormat/wiki/Async-and-Thread-Safety)? It might already answer your questions. If not, let me know the missing parts please.

**Comment by caviyacht** [comment]
@axunonb , I have read that, but I don't think it explains that you cannot use the same instance of SmartFormatter in an async/parallel environment. If you see this original post and my follow-up comment, you can see that using a single SmartFormatter instance will eventually fail to parse; this is due to the internal parser not being created on each parse (see https://github.com/axuno/SmartFormat/blob/423466bb6046710c58d21f849bad029a41bd7428/src/SmartFormat/SmartFormatter.cs#L34), but rather reused. This ends up keeping state (https://github.com/axuno/SmartFormat/blob/423466bb6046710c58d21f849bad029a41bd7428/src/SmartFormat/Core/Parsing/Parser.cs#L231) and I feel like is causing this issue.

Here is another post about the same issue. It doesn't say exactly what they were doing, but it is possible they were reusing the instance (as it seems like you should): https://github.com/axuno/SmartFormat/issues/374

tl;dr, I guess the solution would be to either change the internal parser to make a new one on parse per use (as everything else inside seems to be utilizing pools), or update the documentation saying that an instance cannot be used across threads, which I feel like isn't the goal.

**Comment by axunonb** [comment]
[Version 3.6.0](https://github.com/axuno/SmartFormat/releases) has just been released. Any comments welcome.
